src/Microsoft.DotNet.GenAPI/Shared/CSharpBuilder.cs

```diff
@@ -92,38 +97,91 @@ protected override void Process(AttributeData data)
 
     private void Process(IPropertySymbol ps)
     {
-        _syntaxWriter.WriteProperty(ps.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
-            hasImplementation: !ps.IsAbstract, ps.GetMethod != null, ps.SetMethod != null);
+        _syntaxWriter.WriteProperty(
+            BuildMemberAccessibility(ps),
+            ps.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
+            hasImplementation: !ps.IsAbstract,
+            ps.GetMethod != null,
+            ps.SetMethod != null);
     }
 
     private void Process(IEventSymbol es)
     {
-        _syntaxWriter.WriteEvent(es.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
-            es.AddMethod != null, es.RemoveMethod != null);
+        _syntaxWriter.WriteEvent(
+            BuildMemberAccessibility(es),
+            es.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
+            !es.IsAbstract && es.AddMethod != null,
+            !es.IsAbstract && es.RemoveMethod != null);
     }
 
     private void Process(IMethodSymbol ms)
     {
-        _syntaxWriter.WriteMethod(ms.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
+        _syntaxWriter.WriteMethod(
+            BuildMemberAccessibility(ms),
+            ms.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat),
             hasImplementation: !ms.IsAbstract);
     }
 
     private void Process(IFieldSymbol field)
     {
-       _syntaxWriter.WriteField(field.ToDisplayString(new(
-            memberOptions: SymbolDisplayMemberOptions.IncludeConstantValue)));
+        if (field.ContainingType.TypeKind == TypeKind.Enum)
+        {
+            _syntaxWriter.WriteEnumField(
+                field.ToDisplayString(new(memberOptions: 
+                    SymbolDisplayMemberOptions.IncludeConstantValue)));
+        }
+        else
+        {
+            _syntaxWriter.WriteField(
+                BuildMemberAccessibility(field),
+                field.ToDisplayString(AssemblySymbolDisplayFormats.MemberDisplayFormat));
+        }
+    }
+
+    private bool NeedsAccessibility(ISymbol symbol)
+    {
+        return symbol switch
+        {
+            INamespaceSymbol => false,
+            INamedTypeSymbol => false,
+            IFieldSymbol fs => fs.ContainingType.TypeKind != TypeKind.Enum,
+            IMethodSymbol ms =>
+                !ms.ExplicitInterfaceImplementations.Any() &&
+                 ms.ContainingType.TypeKind != TypeKind.Interface,
+            IPropertySymbol ps =>
+                !ps.ExplicitInterfaceImplementations.Any() &&
+                 ps.ContainingType.TypeKind != TypeKind.Interface,
+            _ => true
+        };
+    }
```

Comment:
FYI there is also this notation (not sure which one you prefer):

```suggestion
    private bool NeedsAccessibility(ISymbol symbol) => symbol switch
    {
        INamespaceSymbol => false,
        INamedTypeSymbol => false,
        IFieldSymbol fs => fs.ContainingType.TypeKind != TypeKind.Enum,
        IMethodSymbol ms => !ms.ExplicitInterfaceImplementations.Any() && ms.ContainingType.TypeKind != TypeKind.Interface,
        IPropertySymbol ps => !ps.ExplicitInterfaceImplementations.Any() && ps.ContainingType.TypeKind != TypeKind.Interface,
        _ => true
    };
```

Reply:
Any time there's just the `return` statement, you can use this property-based block body